### PR TITLE
Fix old I03 data not processing

### PR DESCRIPTION
**Pull request by ndevenish**
This data did not process, because:
- dxtbx could not find the `_meta.h5` file (it was not linked from the `.nxs`, which was assumed). There is now a fallback that uses pure filename instead of trying to determine it from the main file.
- The `_meta.h5` file didn't have a stored "Detector bit depth" field. This now uses the per-image `bitdepth` field.
- General cleanup/update to pathlib.
- 
Fixes dials/dials#2669.

### CODE REVIEW

File: newsfragments/737.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix case where old I03 nexus data (pre-2020) would fail to process.
```
Comment (graeme-winter) [inline newsfragments/737.bugfix:1]: ... eiger ...

File: src/dxtbx/format/FormatNXmxDLS.py [hunk 2]
```diff
@@ -13,36 +12,63 @@
...
-def get_bit_depth_from_meta(meta_file_name):
+def get_bit_depth_from_meta(meta_file_name: Path) -> int:
     with h5py.File(meta_file_name) as f:
-        return int(f["/_dectris/bit_depth_image"][()])
+        with contextlib.suppress(KeyError):
+            return int(f["/_dectris/bit_depth_image"][()])
 
+        # This might be a very old meta file, that only had 'datatype'?
+        if "datatype" in f:
+            unique_vals = np.unique(f["datatype"])
```
Comment (graeme-winter) [inline src/dxtbx/format/FormatNXmxDLS.py:22]: Sensible, but not sure it is necessary i.e. just return the biggest?

File: src/dxtbx/format/FormatNXmxDLS.py [hunk 3]
```diff
@@ -13,36 +12,63 @@
...
+            if not len(unique_vals) == 1:
+                raise RuntimeError(
```
Comment (graeme-winter) [inline src/dxtbx/format/FormatNXmxDLS.py:24]: `sys.exit()`?

File: src/dxtbx/format/FormatNXmxDLS.py [hunk 4]
```diff
@@ -13,36 +12,63 @@
...
+                    f"Error: Could not determine bit depth from metafile {meta_file_name} (multiple bit depths)"
+                )
+            return np.dtype(unique_vals[0]).itemsize * 8
 
-def find_meta_filename(master_like: Union[str, Path]) -> Union[str, Path]:
+    raise RuntimeError("Cannot determine metafile bit depth")
+
+
+def find_meta_filename(master_like: Path) -> Path:
```
Comment (graeme-winter) [inline src/dxtbx/format/FormatNXmxDLS.py:32]: Does this work with `foo.nxs` with no internal references to `foo_meta.h5`? I can see `foo_master.h5` etc. being easy

File: src/dxtbx/format/FormatNXmxDLS.py [hunk 5]
```diff
@@ -13,36 +12,63 @@
...
-    master_dir = os.path.split(master_like)[0]
+    master_like = Path(master_like)
     with h5py.File(master_like) as f:
         meta_filename = f.visit(_local_visit)
 
-    return os.path.join(master_dir, meta_filename)
+    if meta_filename is None:
+        # Try a fallback of looking for a file named the same but "meta.h5"
+        look_for = master_like.with_name(
+            master_like.stem.removesuffix("_master") + "_meta.h5"
```
Comment (graeme-winter) [inline src/dxtbx/format/FormatNXmxDLS.py:64]: i.e. if `foo.nxs` is passed what does this do?

**Review by graeme-winter — APPROVED** [review]
Approve as-is because would fix the problem as reported, I assume, but looks like there is a gap woth handling `foo.nxs`

File: src/dxtbx/format/FormatNXmxDLS.py (hunk 2 above)
Comment (ndevenish) [reply src/dxtbx/format/FormatNXmxDLS.py:22]: I'm not confident I know the implications of that for handling multi-data-type datasets, and we don't have any test data for it so I don't know if doing that would make things work. I'd rather keep this working on the simple case and address any other failures later.

File: src/dxtbx/format/FormatNXmxDLS.py (hunk 3 above)
Comment (ndevenish) [reply src/dxtbx/format/FormatNXmxDLS.py:24]: https://github.com/cctbx/dxtbx/blob/88c7db30652aa15249e3837de8ff7db1f671caa5/src/dxtbx/format/FormatNXmxDLS.py#L79-L80 is what I was basing this on. This caught _all_ exceptions with a fallback, now it only catches a limited range (which doesn't catch e.g. AttributeError which is what was causing the problem before).

I should at least add a log warning that a fallback is being used.

File: src/dxtbx/format/FormatNXmxDLS.py (hunk 5 above)
Comment (ndevenish) [reply src/dxtbx/format/FormatNXmxDLS.py:64]:
```python
>>> Path("foo.nxs").stem.removesuffix("_master") + "_meta.h5"
'foo_meta.h5'
>>> Path("foo_master.h5").stem.removesuffix("_master") + "_meta.h5"
'foo_meta.h5'
```

File: src/dxtbx/format/FormatNXmxDLS.py (hunk 4 above)
Comment (ndevenish) [reply src/dxtbx/format/FormatNXmxDLS.py:32]: Yes, I tested with David's test case, which was using the `.nxs`. I just also checked with `_master.h5` and that works too.

**Comment by dagewa** [comment]
Thanks, confirmed I can process that data set now 👍

**Pull request merged**